### DISCUSSION TITLE
Circularbuffer vs EEPROM.read/write

### DISCUSSION THREAD

**Discussion by Francois1976** [rlogiacco/CircularBuffer#31] [Q&A] [answered]
Hi, I would like to restore the value of my buffer during my arduino bootup.  Is there a way to do it?

My buffer is define like that and work well:
  CircularBuffer<int, 24> buffer;

I backup like this:
```
   EEPROM.update(0, buffer[0]);
   EEPROM.update(1, buffer[1]);
   EEPROM.update(2, buffer[2]);
```

I try to restore like that on the SETUP part of my program
```
  buffer[0]=EEPROM.read(0);
  buffer[1]=EEPROM.read(1);
  buffer[2]=EEPROM.read(2);
```

but I cannot I received the following error during the compile:

> "lvalue required as left operand of assignment" 

I also tryed some other way to restore but never find the good one....

So for my bad english  ;o)

Thanks!

**Comment by rlogiacco** [comment, marked as answer]
The message is quite obvious, but to put it simple, you are trying to treat the buffer as if it was an array and use the [] to identify an item in the buffer and perform an assignment: that is not going to work.

The [] operator can be used to read data from the buffer, not to perform assignments, that is why you find the [] operator listed in the [_retrieve data_](https://github.com/rlogiacco/CircularBuffer#retrieve-data) section of the readme but **not** in the [_store data_](https://github.com/rlogiacco/CircularBuffer#store-data).

What you can do is replace your
```
// this cannot work because the left operand is not assignable (lvalue)
buffer[0]=EEPROM.read(0);
buffer[1]=EEPROM.read(1);
buffer[2]=EEPROM.read(2);
```
with 
```
// this works because we are appending to the buffer
buffer.push(EEPROM.read(0));
buffer.push(EEPROM.read(1));
buffer.push(EEPROM.read(2));
```
or even better with
```
// read all the 24 positions and restore the buffer contents
for (uint8_t i = 0; i < 24; i++) {
  buffer.push(EEPROM.read(i));
}
```
